File: debop4k-web/src/main/java/debop4k/web/AbstractSpringWebApplicationInitializer.java
package debop4k.web;

import debop4k.core.spring.Profiles;
import kotlin.text.Charsets;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.web.context.WebApplicationContext;
import org.springframework.web.filter.CharacterEncodingFilter;
import org.springframework.web.servlet.support.AbstractAnnotationConfigDispatcherServletInitializer;

import javax.servlet.Filter;

/**
 * Spring MVC 용 Servlet Initializer 입니다.
 * web.xml 을 사용하지 않고, 코드 상에서 정의합니다.
 *
 * @author sunghyouk.bae@gmail.com
 * @since 2015. 9. 13.
 */
@Slf4j
public abstract class AbstractSpringWebApplicationInitializer
    extends AbstractAnnotationConfigDispatcherServletInitializer {

  @Override
  protected String[] getServletMappings() {
    return new String[]{"/"};
  }

  @Override
  protected Filter[] getServletFilters() {
    CharacterEncodingFilter filter = new CharacterEncodingFilter();
    filter.setEncoding(Charsets.UTF_8.toString());

    return new Filter[]{filter};
  }

  @Override
  protected WebApplicationContext createServletApplicationContext() {
    WebApplicationContext webAppContext = super.createServletApplicationContext();
    return setProfiles(webAppContext);
  }

  private WebApplicationContext setProfiles(WebApplicationContext context) {
    ConfigurableApplicationContext ctx = (ConfigurableApplicationContext) context;
    String profile = System.getProperty("profile", Profiles.Local.name()).toLowerCase();

    log.info("환경설정 중 active profile 을 지정합니다. profile={}", profile);

    if (ctx != null) {
      ctx.getEnvironment().setActiveProfiles(profile);
    }
    return (WebApplicationContext) ctx;
  }
}

File: debop4k-web/src/main/kotlin/debop4k/web/AbstractSpringWebApplicationInitializer.kt
/*
 * Copyright (c) 2016. Sunghyouk Bae <sunghyouk.bae@gmail.com>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package debop4k.web

import debop4k.core.loggerOf
import debop4k.spring.config.Profiles
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.web.context.WebApplicationContext
import org.springframework.web.filter.CharacterEncodingFilter
import org.springframework.web.servlet.support.AbstractAnnotationConfigDispatcherServletInitializer
import javax.servlet.Filter

/**
 * Spring MVC 용 Servlet Initializer 입니다.
 * web.xml 을 사용하지 않고, 코드 상에서 정의합니다.
 *
 * @author sunghyouk.bae@gmail.com
 */
abstract class AbstractSpringWebApplicationInitializer : AbstractAnnotationConfigDispatcherServletInitializer() {

  protected val log = loggerOf(javaClass)

  override fun getServletMappings(): Array<out String> {
    return arrayOf<String>()
  }

  override fun getServletFilters(): Array<out Filter> {
    val filter = CharacterEncodingFilter().apply {
      encoding = Charsets.UTF_8.toString()
    }
    return arrayOf(filter)
  }

  override fun createServletApplicationContext(): WebApplicationContext {
    val context = super.createServletApplicationContext()
    return setProfiles(context)
  }

  private fun setProfiles(context: WebApplicationContext): WebApplicationContext {
    val ctx = context as? ConfigurableApplicationContext
    val profile = System.getProperty("profile", Profiles.Local.name).toLowerCase()

    log.info("환경설정 중 active profile을 지정합니다. active profile={}", profile)

    if (ctx != null)
      ctx.environment.setActiveProfiles(profile)

    return ctx as WebApplicationContext
  }
}